File: src/rust_solutions/1358_number-of-substrings-containing-all-three-characters.rs
#[allow(unused)]
use crate::stubs::*;

struct Solution;

// start_submission
impl Solution {
    pub fn number_of_substrings(s: String) -> i32 {
        let mut seen = [0; 3];

        let mut left_ptr = 0;
        let mut right_ptr = 0;

        let mut count = 0;

        while right_ptr < s.len() {
            seen[(s.as_bytes()[right_ptr] - b'a') as usize] += 1;

            while seen[0] > 0 && seen[1] > 0 && seen[2] > 0 {
                count += (s.len() - right_ptr) as i32;

                seen[(s.as_bytes()[left_ptr] - b'a') as usize] -= 1;

                left_ptr += 1;
            }

            right_ptr += 1;
        }

        count
    }
}
// end_submission

File: src/cpp_solutions/1358_number-of-substrings-containing-all-three-characters.cpp
#include <bits/stdc++.h>

using namespace std;

// start_submission
class Solution {
public:
    int numberOfSubstrings(string s) {
        int n = s.size();

        unordered_map<char, int> m;

        int ans = 0;

        int l = 0;

        for (int r = 0; r < n; ++r) {
            ++m[s[r]];

            while (l < r && m.size() == 3 && m[s[l]] > 1) {
                --m[s[l]];
                ++l;
            }

            if (m.size() == 3) {
                ans += l + 1;
            }
        }

        return ans;
    }
};
// end_submission
